transaction/core/src/memo.rs

```diff
@@ -0,0 +1,218 @@
+    derive_into_vec_from_repr_bytes, derive_prost_message_from_repr_bytes,
+    derive_repr_bytes_from_as_ref_and_try_from, derive_serde_from_repr_bytes,
+};
+use sha2::Sha512;
+
+/// An encrypted memo, which can be decrypted by the recipient of a TxOut.
+#[derive(Clone, Copy, Default, Debug, Eq, Hash, Digestible, Ord, PartialEq, PartialOrd)]
+pub struct EncryptedMemo(GenericArray<u8, U46>);
+
+impl AsRef<[u8]> for EncryptedMemo {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_slice()
+    }
+}
+
+impl AsRef<GenericArray<u8, U46>> for EncryptedMemo {
+    fn as_ref(&self) -> &GenericArray<u8, U46> {
+        &self.0
+    }
+}
+
+impl From<EncryptedMemo> for GenericArray<u8, U46> {
+    fn from(src: EncryptedMemo) -> Self {
+        src.0
+    }
+}
+
+impl From<GenericArray<u8, U46>> for EncryptedMemo {
+    fn from(src: GenericArray<u8, U46>) -> Self {
+        Self(src)
+    }
+}
+
+impl TryFrom<&[u8]> for EncryptedMemo {
+    type Error = LengthError;
+    fn try_from(src: &[u8]) -> Result<EncryptedMemo, Self::Error> {
+        if src.len() == 46 {
+            Ok(Self(*GenericArray::from_slice(src)))
+        } else {
+            Err(LengthError::BadLength(src.len()))
+        }
+    }
+}
+
+derive_repr_bytes_from_as_ref_and_try_from!(EncryptedMemo, U46);
+derive_into_vec_from_repr_bytes!(EncryptedMemo);
+derive_serde_from_repr_bytes!(EncryptedMemo);
+derive_prost_message_from_repr_bytes!(EncryptedMemo);
+
+impl EncryptedMemo {
+    /// Helper to ease syntax when decrypting
+    ///
+    /// The shared-secret is expected to be the TxOut shared secret of the TxOut
+    /// that this memo is associated to.
+    pub fn decrypt(&self, shared_secret: &RistrettoPublic) -> MemoPayload {
+        MemoPayload::decrypt_from(self, shared_secret)
+    }
+}
+
+/// A plaintext memo payload, with accessors to easily access the memo type
+/// bytes and memo data bytes. High-level memo objects should be convertible
+/// to MemoPayload. Deserialization, across all high-level memo types, is
+/// done in mc-transaction-std crate.
+///
+/// Note that a memo payload may be invalid / uninterpretable, or refer to new
+/// memo types that have been introduced at a later date.
+#[derive(Clone, Copy, Default, Debug, Eq, Digestible, Ord, PartialEq, PartialOrd)]
+pub struct MemoPayload(GenericArray<u8, U46>);
+
+impl MemoPayload {
+    /// Create a new memo payload from given type bytes and data bytes
+    pub fn new(memo_type: [u8; 2], memo_data: [u8; 44]) -> Self {
+        let mut result = Self::default();
+        result.0[0..2].copy_from_slice(&memo_type);
+        result.0[2..46].copy_from_slice(&memo_data);
+        result
+    }
+
+    /// Get the memo type bytes (two bytes)
+    pub fn get_memo_type(&self) -> &[u8; 2] {
+        self.0.as_slice()[0..2].try_into().unwrap()
+    }
+
+    /// Get the memo data bytes (fourty-four bytes)
+    pub fn get_memo_data(&self) -> &[u8; 44] {
+        self.0.as_slice()[2..46].try_into().unwrap()
```

Comment:
```suggestion
        self.0.as_slice()[2..46].try_into().expect("Could not convert 44-byte slice to a 44-byte arrayref");
```

Reply:
i feel like, this is just too verbose of a thing for me to want to type this everywhere we have a conversion like this

if you really aren't okay with `unwrap` maybe we can do `.expect("length mismatch")` ?